core/context-strategy/autoscale-context.ts

```diff
@@ -106,87 +132,532 @@ export class PreferredGroupPrimaryElection implements PrimaryElectionStrategy {
+
     /**
-     * Force the target vm to go into 'out-of-sync' state. Autoscale will stop accepting its
-     * heartbeat sync request.
-     * @returns {Promise} void
+     * This method will run the classic primary election that does not facilitate the device
+     * sync info provided by each VM
+     * @returns {PrimaryElectionStrategyResult} election result
      */
-    forceOutOfSync(): Promise<boolean>;
-    readonly targetHealthCheckRecord: HealthCheckRecord | null;
-    readonly healthCheckResult: HealthCheckResult;
-    readonly healthCheckResultDetail: HealthCheckResultDetail;
-    readonly targetVmFirstHeartbeat: boolean;
+    private runClassicElectionMethod(): Promise<PrimaryElectionStrategyResult> {
+        return super.run();
+    }
+
+    /**
+     * This method will run the primary election that makes uses of the device sync info
+     * provided by each VM. The sync info will be kept in the healthcheck record of each VM
+     * @returns {PrimaryElectionStrategyResult} election result
+     */
+    private async runDeviceSyncInfoElectionMethod(): Promise<PrimaryElectionStrategyResult> {
+        const settings = await this.platform.getSettings();
+        // get the primary scaling group
+        const settingGroupName = settings.get(AutoscaleSetting.PrimaryScalingGroupName).value;
+        // get all monitored vm health check records and filter the healthy ones
+        const allHealthCheckRecords: HealthCheckRecord[] = (
+            await this.platform.listHealthCheckRecord()
+        )
+            // exclude unhealthy vm from election
+            .filter(rec => rec.healthy)
+            // exclude those are not in the preferred scaling group for primary election
+            .filter(rec => rec.scalingGroupName === settingGroupName);
+
+        // if no vm healthcheck record available, end election
+        if (allHealthCheckRecords.length === 0) {
+            this.proxy.log(
+                "There isn't any healthy vm in the preferred scaling group for primary election." +
+                    'Primary election not started.',
+                LogLevel.Warn
+            );
+            this.res.newPrimary = null;
+            this.res.newPrimaryRecord = null;
+            return PrimaryElectionStrategyResult.SkipAndContinue;
+        }
+
+        // scores
+        const scores: Map<string, number> = new Map();
+        allHealthCheckRecords.forEach(rec => {
+            scores.set(rec.vmId, 0);
+        });
+        for (const rec of allHealthCheckRecords) {
+            let points = scores.get(rec.vmId);
+            // apply weighting methods
+            points += this.weightMethod1(rec);
+            points += this.weightMethod2(rec, allHealthCheckRecords);
+            points += this.weightMethod3(rec, allHealthCheckRecords);
+            points += this.weightMethod4(rec);
+            points += this.weightMethod5(rec, allHealthCheckRecords);
+            points += this.weightMethod6(rec, allHealthCheckRecords);
```

Comment:
I think the weight methods should have descriptive names instead of 1 2 3 4 5 6....

haveChecksumWeight() + sharedChecksumWeight() + groupedChecksumWeight() + ...  etc

Reply:
Some weigh methods have mixed and complicated conditions which are hard to have a self-descriptive name so I gave up naming them in that way. I prefer to keep their naming well-formatted and with a better jsdoc. It also does the job. Any good IDE could display the information properly.

@jamie-pate  Let me know whether the self-descriptive naming is still strongly desired to apply here in this case or not?

Reply:
fixed